File: c++/135. Candy.cpp
class Solution {
public:
    /**
     * @param ratings Children's ratings
     * @return the minimum candies you must give
     */
    int candy(vector<int>& ratings) {
        // Write your code here
        int ls = ratings.size();
        if (ls == 0) return 0;
        if (ls == 1) return 1;
        vector<int> candy(ls, 1);
        int res = 0;
        for (int i = 0; i < ls - 1; i ++) {
            if (ratings[i+1] > ratings[i])
                candy[i+1] = candy[i] + 1;
        }
        for (int i = ls-1; i > 0; i--) {
            if (ratings[i-1] > ratings[i])
                candy[i-1] = max(candy[i-1], candy[i]+1);
        }
        for (int i = 0; i < ls; i ++)
            res += candy[i];
        
        return res;
    }
};

File: python/135. Candy.py
class Solution(object):
    def candy(self, ratings):
        """
        :type ratings: List[int]
        :rtype: int
        """
        n = len(ratings)
        candies = [1]*n
        for i in range(1, n):
            if ratings[i] > ratings[i-1]:
                candies[i] = candies[i-1]+1
        
        for i in range(n-1, 0, -1):
            if ratings[i-1] > ratings[i]:
                candies[i-1] = max(candies[i-1], candies[i]+1)
        return sum(candies)

# 思路： 先从左向右遍历，确定右边元素的candy， 再从右向左遍历，确定左边元素的candy        